pkg/tests/testsupport/metric_loader.go

```diff
@@ -0,0 +1,245 @@
+	labels := i.labelsCache[blockSample.lblsHash]
+	protoLabels := make([]prompb.Label, len(labels))
+	for i, l := range labels {
+		protoLabels[i] = prompb.Label{Name: l.Name, Value: l.Value}
+	}
+	sample := prompb.Sample{
+		Value:     blockSample.val,
+		Timestamp: blockSample.ts,
+	}
+	ts := prompb.TimeSeries{
+		Labels:               protoLabels,
+		Samples:              []prompb.Sample{sample},
+		Exemplars:            []prompb.Exemplar{},
+		XXX_NoUnkeyedLiteral: struct{}{},
+		XXX_unrecognized:     []byte{},
+		XXX_sizecache:        0,
+	}
+
+	return timeSeries{blockSample.lblsHash, ts}
+}
+
+func NewPromLoader(dataDir string) (PromLoader, error) {
+	db, err := tsdb.OpenDBReadOnly(dataDir, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error starting Prometheus TSDB in read-only: %v", err)
+	}
+	blocks, err := db.Blocks()
+	if err != nil {
+		return nil, fmt.Errorf("error loading data blocks: %v", err)
+	}
+	return &promLoader{db: db, blocks: blocks}, nil
+}
+
+func (loader *promLoader) Iterator() PromIterator {
+	return &promIterator{blocks: loader.blocks, curSampleIdx: -1, curBlockIdx: -1}
+}
+
+func (loader *promLoader) Close() error {
+	return loader.db.Close()
+}
+
+type IngestFunc func(context.Context, *prompb.WriteRequest) (uint64, uint64, error)
+
+type sampleIngestor struct {
+	samples   PromIterator
+	shards    []chan prompb.TimeSeries
+	batchSize int
+	rate      *rate.Limiter
+}
+
+// NewSampleIngestor creates sampleIngestor responsible for splitting samples into shards and sending them down the ingestion path
+// zero sampleRate means no limits
+func NewSampleIngestor(shards int, samples PromIterator, batchSize int, sampleRate int) *sampleIngestor {
+	var r *rate.Limiter
+	if sampleRate != 0 {
+		r = rate.NewLimiter(rate.Limit(sampleRate), 0)
+	}
+	sampleLoader := sampleIngestor{
+		samples:   samples,
+		shards:    make([]chan prompb.TimeSeries, shards),
+		batchSize: batchSize,
+		rate:      r,
+	}
+	for i := 0; i < shards; i++ {
+		sampleLoader.shards[i] = make(chan prompb.TimeSeries, 2*batchSize)
+	}
+	return &sampleLoader
+}
+
+func (si *sampleIngestor) Run(ingest IngestFunc) {
+	si.shardSamples()
+	si.ingestSamples(ingest)
+}
+
+func (si *sampleIngestor) shardSamples() {
+	go func() {
+		defer func() {
+			for _, shard := range si.shards {
+				close(shard)
+			}
+		}()
+		for si.samples.Next() {
+			sample := si.samples.Get()
+			shardIdx := sample.seriesHash % uint64(len(si.shards))
+			si.shards[shardIdx] <- sample.Val
+			if si.rate != nil {
+				if err := si.rate.Wait(context.Background()); err != nil {
+					log.Error(err)
+				}
+			}
+		}
+	}()
+}
+
+func (si *sampleIngestor) ingestSamples(ingest IngestFunc) {
+	var wg sync.WaitGroup
+	for i := 0; i < len(si.shards); i++ {
+		go func(shard int) {
+			defer func() {
+				wg.Done()
+			}()
+			var req prompb.WriteRequest = prompb.WriteRequest{Timeseries: make([]prompb.TimeSeries, si.batchSize)}
+			counter := 0
+			for sample := range si.shards[shard] {
```

Comment:
```suggestion
			for ts := range si.shards[shard] {
```
`sample` name is very confusing. This is samples + series, which is proper ts